### PR TITLE
[test_nn] split hooks test from test_nn

### PR DESCRIPTION
**Pull request by kshitij12345**
Ref: https://github.com/pytorch/pytorch/issues/63085

Note: Doesn't need corresponding XLA PR as the migrated tests were not run on XLA (as they weren't in TestNNDeviceType).

### CODE REVIEW

File: test/nn/test_module_hooks.py [hunk 1]
```diff
@@ -192,5 +202,955 @@ def test_mixed_hooks(self):
...
+        def forward_hook(m, input, output):
+            return -output
+
+        nn.modules.module.register_module_forward_pre_hook(forward_pre_hook)
+        nn.modules.module.register_module_forward_hook(forward_hook)
+        output = module(input)
+        expected_res = -torch.sigmoid(torch.nn.functional.relu(input))
+        self.assertEqual(output, expected_res)
+        output.backward(torch.ones(5, 5) * 2, retain_graph=True)
+        mask = (input > 0)
```
Comment (kshitij12345) [inline test/nn/test_module_hooks.py:557]: Used to be `mask = (input > 0).double()` as `test_nn` defaults to `torch.double` dtype.

File: test/nn/test_module_hooks.py [hunk 2]
```diff
@@ -192,5 +202,955 @@ def test_mixed_hooks(self):
...
+        def forward_hook(m, input, output):
+            return -output
+
+        module.register_forward_pre_hook(forward_pre_hook)
+        module.register_forward_hook(forward_hook)
+        output = module(input)
+        expected_res = -torch.nn.functional.sigmoid(torch.nn.functional.relu(input))
+        self.assertEqual(output, expected_res)
+        output.backward(torch.ones(5, 5) * 2, retain_graph=True)
+        mask = (input > 0)
```
Comment (kshitij12345) [inline test/nn/test_module_hooks.py:1101]: Used to be mask = (input > 0).double() as test_nn defaults to torch.double dtype.

File: test/nn/test_module_hooks.py [hunk 3]
```diff
@@ -192,5 +202,958 @@ def test_mixed_hooks(self):
         self.assertEqual(fired_hooks, [0, 1, 2, 3, 0, 1, 2, 3])
 
 
+def _hook_to_pickle(*args, **kwargs):
+    pass
+
+class TestStateDictHooks(TestCase):
```
Comment (albanD) [inline test/nn/test_module_hooks.py:208]: Missing class attributes? :D

Comment (albanD) [reply test/nn/test_module_hooks.py:208]: Awesome!
This will be so much easier to read after this :D

**Review by albanD — APPROVED** [review]
